viewing_party/party.py

```diff
@@ -1,23 +1,266 @@
+    '''
+    insert movie (in dict) in watchlist (in list) in user_data (in dict)
+    exception: if no watchlist (poor baby), create an empty one.
+    '''
+
+    if not user_data["watchlist"]:
+        user_data["watchlist"] = [] 
+
+    user_data["watchlist"].append(movie)
+
+    return user_data
+
+def watch_movie(user_data, title):
+    '''
+    remove movie from watchlist
+    insert movie in watched
+    '''
+    
+    movies_watchlist = user_data["watchlist"] # in list
+
+    for movie in movies_watchlist:
+        if title in movie["title"]:
+            # a removed & added list is reflected to user_data dict because dict is mutable
+            movies_watchlist.remove(movie) 
+            add_to_watched(user_data, movie)
+
+    return user_data
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    '''
+    (1) extract each movie from a list and get each rating
+    (2) sum up rating during loop
+    exception: empty watched (poor!), return 0.0
+    '''
+
+    watched_list = user_data["watched"]
+    rating_sum = 0.0
+    n = len(watched_list)
+    if n == 0:
+        rating_avg = 0.0
+    else:
+        for movie in watched_list:
+            rating = movie["rating"]
+            rating_sum += rating
+        rating_avg = rating_sum/n
+    return rating_avg
+
+def get_most_watched_genre(user_data):
+    '''
+    (1) extract each movie from a list and get each genre
+    (2-0) genres in dict
+    (2-1) new genre: create key & set value 1 
+    (2-2) ex-genre: value +1 
+    (3) find key corresponing max value ! assume no multiple keys :p
+    exception: empty watched (poor!), return None
+    '''
+
+    watched_list = user_data["watched"]
+    
+    if watched_list == []:
+        return None
+    else:
+        genres = {}
+        for movie in watched_list:
+            genre = movie["genre"] 
+            if not genre in genres:
+                genres[genre] = 1
+            else:
+                genres[genre] += 1 
+        most_genre = max(genres, key=genres.get)
+        return most_genre
+
+# -----------------------------------------
+# ------------- Helper Function -----------
+# -----------------------------------------
+def get_value_list_with_same_key_in_dict_list(dict_list, key, value_list):
+    '''
+    get a list of values with the same key from a list of dictionaries
+    value_list: input & output
+    ! does not return value_list as the function can be called recursively in for-loop
+    '''
+
+    for a_dict in dict_list:
+        same_key_value = a_dict[key]
+        # assume ok for redundant as it might have a different details in other keys
+        value_list.append(same_key_value)
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
+def get_unique_watched(user_data):

```

Comment:
Good job! Sets are a great way to remove duplicates and check the difference between two lists. Unfortunately, dictionaries can't be put in a set. In this case, it may be easier to use a for loop to see if a dictionary is in a list. For example, this is a potential solution:
```
for movie in user_movies:
        if movie not in friend_movies:
            user_unique_movie_titles.append(movie)
```
where `user_movies` is the user's movies and `friend_movies` is the list of all of the movies from all of the friends. This comment applies for the next function as well.